### PR TITLE
feat(APIv2): RHINENG-2875 implement tailorings endpoint

### PR DESCRIPTION
**Pull request by romanblanco**
## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [x] General Coding Practices

### CODE REVIEW

File: app/policies/v2/tailoring_policy.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  # Only show tailoring in our user account
+  class Scope < V2::ApplicationPolicy::Scope
+    def resolve
+      resolve scope.where('1=0') if user&.account_id.blank?
+
+      scope.where
+           .associated(:policy)
+           .where(policy: { account_id: user.account_id })
+    end
+  end
```
Comment (skateman) [inline app/policies/v2/tailoring_policy.rb:28]: Missing spec for tailoring_policy

File: spec/controllers/v2/concerns/rendering_spec.rb [hunk 2]
```diff
@@ -50,7 +50,9 @@
     expect(response_body_data).to match_array(collection)
 
     response_body_data.each do |item|
-      expect(item.keys.count).to eq(attributes.keys.count + 2)
+      default_attributes = %i[id type].freeze
+      expected_attributes = default_attributes + attributes.keys
+      expect(item.keys.count).to eq(expected_attributes.count)
```
Comment (skateman) [inline spec/controllers/v2/concerns/rendering_spec.rb:55]: I am not saying this is wrong, just need some explanation for this

File: spec/controllers/v2/tailorings_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        FactoryBot.create(
+          :v2_tailoring,
+          policy: parent,
+          profile: canonical_profile,
+          os_minor_version: 8
+        )
+      end
+
+      it 'returns XCCDF tailoring file' do
+        header = JSON.parse(Base64.decode64(current_user.account.identity_header.raw))
```
Comment (skateman) [inline spec/controllers/v2/tailorings_controller_spec.rb:126]: If you don't use `raw` you might be able to avoid the decoding

File: spec/factories/rule.rb [hunk 4]
```diff
@@ -16,6 +16,20 @@
       profile_id { nil }
     end
 
+    trait :with_group do
+      rule_group { association :v2_rule_group, security_guide: security_guide }
+    end
+
+    trait :with_group_hierarchy do
```
Comment (skateman) [inline spec/factories/rule.rb:23]: These can be eliminated as I explained last week

File: spec/controllers/v2/concerns/rendering_spec.rb (hunk 2 above)
Comment (romanblanco) [reply spec/controllers/v2/concerns/rendering_spec.rb:55]: It could be simplified to

```suggestion
      expect(item.keys.count).to eq((attributes.keys + %i[id type]).count)
```

My main focus is to get rid of the *magic* number `2` and clarify what it stands for.

Comment (skateman) [reply spec/controllers/v2/concerns/rendering_spec.rb:55]: This is outside of the scope of this PR

Comment (romanblanco) [reply spec/controllers/v2/concerns/rendering_spec.rb:55]: Ack

File: app/models/v2/tailoring.rb [hunk 5]
```diff
@@ -3,19 +3,61 @@
...
+    def value_overrides
+      attributes['tailoring__value_overrides'] || profile.value_overrides
+    end
+
+    def tailoring_rule_group_ref_ids
+      base = V2::RuleGroup.where(id: rules.select(:rule_group_id))
+      base
+        .or(V2::RuleGroup.where(id: base.select(GROUP_ANCESTRY_IDS)))
+        .order(:precedence)
+        .pluck(:ref_id)
```
Comment (skateman) [inline app/models/v2/tailoring.rb:60]:
```suggestion
      base.or(V2::RuleGroup.where(id: base.select(GROUP_ANCESTRY_IDS)))
              .order(:precedence)
              .pluck(:ref_id)
```

File: spec/controllers/v2/tailorings_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        header['identity']['auth_type'] = Insights::Api::Common::IdentityHeader::CERT_AUTH
+        request.headers['X-RH-IDENTITY'] = Base64.encode64(header.to_json)
+
+        get :tailoring_file,
+            params: extra_params.reject { |_, ep| ep.is_a?(ActiveRecord::Base) }.merge(parents: [:policy])
```
Comment (skateman) [inline spec/controllers/v2/tailorings_controller_spec.rb:131]: 🤯 how this ended up here? And why is it so complicated?

File: spec/controllers/v2/tailorings_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        expect(response).to have_http_status :ok
+        expect(response.headers['Content-Type']).to eq('application/xml')
```
Comment (skateman) [inline spec/controllers/v2/tailorings_controller_spec.rb:133]: Is the tailoring content actually tested? Also what about value definitions?

File: spec/factories/rule.rb (hunk 4 above)
Comment (romanblanco) [reply spec/factories/rule.rb:23]: That can be done outside of this PR.

File: spec/controllers/v2/tailorings_controller_spec.rb (hunk 7 above)
Comment (romanblanco) [reply spec/controllers/v2/tailorings_controller_spec.rb:133]: Content is not tested. I've only verified it manually.

File: spec/factories/rule.rb (hunk 4 above)
Comment (skateman) [reply spec/factories/rule.rb:23]: You are introducing complexity here that you want to remove later?

File: spec/controllers/v2/tailorings_controller_spec.rb (hunk 7 above)
Comment (skateman) [reply spec/controllers/v2/tailorings_controller_spec.rb:133]: Then what's the test for? And the factory around it?

File: app/models/v2/rule_group.rb [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+module V2
+  # Stores information about Rule Groups. This (eventually) comes from SCAP import.
+  class RuleGroup < ApplicationRecord
+    # FIXME: clean up after the remodel
+    self.table_name = :v2_rule_groups
+    self.primary_key = :id
+
+    belongs_to :security_guide
+
+    has_ancestry(primary_key_format: %r{\A[\w\-]+(\/[\w\-]+)*\z})
```
Comment (skateman) [inline app/models/v2/rule_group.rb:12]: Consistency:
```suggestion
    has_ancestry primary_key_format: %r{\A[\w\-]+(\/[\w\-]+)*\z}
```

File: app/models/v2/tailoring.rb [hunk 9]
```diff
@@ -3,19 +3,60 @@
...
-    validates :os_minor_version, numericality: { greater_than_or_equal_to: 0 }
+    validates :os_minor_version, numericality: { greater_than_or_equal_to: 0 }, uniqueness: { scope: :policy }
+
+    def os_major_version
+      attributes['security_guide__os_major_version'] || security_guide.os_major_version
+    end
+
+    def value_overrides
+      attributes['tailoring__value_overrides'] || profile.value_overrides
+    end
```
Comment (skateman) [inline app/models/v2/tailoring.rb:53]: 1. there is inconsistency between the cached (left) and the non-cached (right) evaluation
2. there is no `derived_attribute` for this in the serializer
3. the left side doesn't even make sense as it points to an own column
4. not sure if you want to ever retrieve what is returned by the right side

File: app/policies/v2/tailoring_policy.rb [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    end
+
+    def show?
+      match_account?
+    end
+
+    def tailoring_file?
+      user.cert_authenticated?
+    end
+    # Only show tailoring in our user account
```
Comment (skateman) [inline app/policies/v2/tailoring_policy.rb:17]: Missing newline
```suggestion
    end
    
    # Only show tailoring in our user account
```

File: app/services/v2/xccdf_tailoring_file.rb [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+      xml[XCCDF].benchmark(
+        id: @tailoring.security_guide.ref_id,
+        version: @tailoring.security_guide.version,
+        href: "ssg-rhel#{@tailoring.security_guide.os_major_version}-ds.xml"
+      )
+    end
+
+    def profile_builder(xml)
+      xml[XCCDF].Profile(id: @tailoring.policy.ref_id,
+                         extends: @tailoring.profile.ref_id) do
```
Comment (skateman) [inline app/services/v2/xccdf_tailoring_file.rb:39]:
```suggestion
      xml[XCCDF].Profile(id: @tailoring.policy.ref_id, extends: @tailoring.profile.ref_id) do
```

File: app/services/v2/xccdf_tailoring_file.rb [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+      rule_group_selections_builder(xml)
+      value_builder(xml)
+    end
+
+    def create_builder
+      validate_input
+      Nokogiri::XML::Builder.new(encoding: 'UTF-8') do |xml|
+        xml[XCCDF].Tailoring('xmlns:xccdf' =>
+                             'http://checklists.nist.gov/xccdf/1.2',
+                             id: 'xccdf_csfr-compliance_tailoring_default') do
```
Comment (skateman) [inline app/services/v2/xccdf_tailoring_file.rb:80]: This is very hard to read 😞

File: app/services/v2/xccdf_tailoring_file.rb [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+    def validate_input
+      handle_missing_profile
+      handle_missing_rules
+    end
+
+    def handle_missing_rules
+      missing_rules = @rule_ref_ids -
+                      @tailoring.profile.rules.map(&:ref_id) -
+                      @tailoring.security_guide.rules.map(&:ref_id)
+      # TOOD:                    .where(ref_id: @rule_ref_ids).pluck(:ref_id)
```
Comment (skateman) [inline app/services/v2/xccdf_tailoring_file.rb:95]:
```suggestion
      # TODO:                    .where(ref_id: @rule_ref_ids).pluck(:ref_id)
```
And what it even means?

File: config/routes.rb [hunk 14]
```diff
@@ -34,8 +34,11 @@ def draw_routes(prefix)
             end
           end
 
-          resources :policies, except: [:new, :edit]
-
+          resources :policies, except: [:new, :edit] do
+            resources :tailorings, only: [:index, :show], parents: [:policy] do
+              get 'tailoring_file', on: :member
```
Comment (skateman) [inline config/routes.rb:39]:
```suggestion
              get :tailoring_file, on: :member
```

File: spec/controllers/v2/tailorings_controller_spec.rb [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+
+  before do
+    request.headers['X-RH-IDENTITY'] = current_user.account.identity_header.raw
+    allow(StrongerParameters::InvalidValue).to receive(:new) { |value, _| value.to_sym }
+    allow(controller).to receive(:rbac_allowed?).and_return(rbac_allowed?)
+  end
+
+  context '/policies/:id/tailorings' do
+    describe 'GET index' do
+      let(:canonical_profiles) do
```
Comment (skateman) [inline spec/controllers/v2/tailorings_controller_spec.rb:26]:
```suggestion
      # Build a set of canonical profiles for various OS minor versions across benchmarks
      let(:canonical_profiles) do
```

File: spec/controllers/v2/tailorings_controller_spec.rb [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+          account: current_user.account,
+          profile: FactoryBot.create(:v2_profile)
+        ).id)
+      end
+
+      it_behaves_like 'individual', :policy
+    end
+  end
+
+  context '/policies/:id/tailorings/:id/tailoring_file' do
```
Comment (skateman) [inline spec/controllers/v2/tailorings_controller_spec.rb:96]: I expect the following contexts underneath:
- With random rules and values
- With default rules and values
- Trying to access a tailoring file with a mismatching set of IDs (404)
- Trying to access a tailoring file under a policy of a different user (404)

The last 2 cases are covered in the show/index automatically by the shared example + `TailoringPolicySpec`.

File: spec/factories/rule.rb (hunk 4 above)
Comment (skateman) [reply spec/factories/rule.rb:23]: Please eliminate the unnecessary complexity that you are introducing 🙏

File: spec/fixtures/files/sortable/tailorings_controller.yaml [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+  - :factory: :v2_tailoring
+    :os_minor_version: 0
+    :policy_id: ${policy_id}
+    :ref_id: ${policy.ref_id}
+
+  - :factory: :v2_tailoring
+    :os_minor_version: 1
+    :policy_id: ${policy_id}
+    :ref_id: ${policy.ref_id}
+
```
Comment (skateman) [inline spec/fixtures/files/sortable/tailorings_controller.yaml:11]: Another one with `os_minor_version: 1` that has to be reflected in the sorting to test if the secondary sorting works well using IDs.

File: spec/policies/v2/tailoring_policy_spec.rb [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+  end
+
+  it 'allows displaying entities related to current user' do
+    expect(Pundit.policy_scope(user, V2::Tailoring).map(&:id)).to eq([item.id])
+  end
+
+  it 'authorizes the index and show actions' do
+    expect(Pundit.authorize(user, item, :index?)).to be_truthy
+    expect(Pundit.authorize(user, item, :show?)).to be_truthy
+  end
```
Comment (skateman) [inline spec/policies/v2/tailoring_policy_spec.rb:48]: There is no test for accessing the wrong tailoring and getting not authorized.

**Review by skateman — CHANGES_REQUESTED** [review]

File: app/services/v2/xccdf_tailoring_file.rb (hunk 13 above)
Comment (romanblanco) [reply app/services/v2/xccdf_tailoring_file.rb:95]: Forgot to remove it

File: spec/fixtures/files/sortable/tailorings_controller.yaml (hunk 17 above)
Comment (romanblanco) [reply spec/fixtures/files/sortable/tailorings_controller.yaml:11]: There can be only one tailoring for each `os_minor_version` under a policy.

File: config/routes.rb [hunk 19]
```diff
@@ -11,7 +11,7 @@ def draw_routes(prefix)
           resources :business_objectives, only: [:index, :show]
           resources :profiles do
             member do
-              get 'tailoring_file'
+              get :tailoring_file
```
Comment (skateman) [inline config/routes.rb:14]: Why? This is APIv1, out of context...

File: app/controllers/v2/tailorings_controller.rb [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+
+    def tailoring
+      @tailoring ||= authorize(expand_resource.find(permitted_params[:id]))
+    end
+
+    def tailored?
+      added_rules = tailoring.rules.order(:precedence) - tailoring.profile.rules
+      removed_rules = tailoring.profile.rules - tailoring.rules.order(:precedence)
+      (added_rules + removed_rules).any?
+    end
```
Comment (skateman) [inline app/controllers/v2/tailorings_controller.rb:42]: This needs a `TODO` comment to check its performance. I believe it is slow 😞 for now it can stay as it is likely copied from the old world, but we need to do it faster.

File: app/models/v2/tailoring.rb [hunk 21]
```diff
@@ -3,19 +3,56 @@
...
     validates :policy, presence: true
     validates :profile, presence: true
-    validates :os_minor_version, numericality: { greater_than_or_equal_to: 0 }
+    validates :os_minor_version, numericality: { greater_than_or_equal_to: 0 }, uniqueness: { scope: :policy }
+
+    def os_major_version
+      attributes['security_guide__os_major_version'] || security_guide.os_major_version
+    end
+
+    def tailoring_rule_group_ref_ids
```
Comment (skateman) [inline app/models/v2/tailoring.rb:51]: Not sure if the `tailoring_` prefix is still necessary...

File: spec/controllers/v2/tailorings_controller_spec.rb [hunk 22]
```diff
@@ -0,0 +1,253 @@
...
+    describe 'GET tailoring_file with default rules and values' do
```
Comment (skateman) [inline spec/controllers/v2/tailorings_controller_spec.rb:98]:
```suggestion
    context 'with default rules and values' do
```

File: spec/controllers/v2/tailorings_controller_spec.rb [hunk 23]
```diff
@@ -0,0 +1,253 @@
...
+        header['identity']['auth_type'] = Insights::Api::Common::IdentityHeader::CERT_AUTH
+        request.headers['X-RH-IDENTITY'] = Base64.encode64(header.to_json)
+
+        get :tailoring_file, params: extra_params.merge(parents: [:policy])
+
+        expect(response).to have_http_status :no_content
+      end
+    end
+
+    describe 'GET tailoring_file with randomly distributed rules' do
```
Comment (skateman) [inline spec/controllers/v2/tailorings_controller_spec.rb:139]:
```suggestion
    context 'with randomly distributed rules' do
```

File: spec/controllers/v2/tailorings_controller_spec.rb [hunk 24]
```diff
@@ -0,0 +1,253 @@
...
+        expect(response.headers['Content-Type']).to eq('application/xml')
+        tailoring_rules.each do |rule|
+          expect(tailored_rules).to include(rule)
+          expect(tailored_rules).to include(V2::Rule.find_by(ref_id: rule).rule_group.ref_id)
+        end
+        expect(tailoring_values).to eq(tailored_values)
+      end
+    end
+
+    describe 'GET tailoring_file with unauthorized policy' do
```
Comment (skateman) [inline spec/controllers/v2/tailorings_controller_spec.rb:194]:
```suggestion
    context 'with unauthorized policy' do
```

File: spec/factories/profile.rb [hunk 25]
```diff
@@ -20,5 +20,38 @@
         association(:profile_os_minor_version, os_minor_version: os_minor_version)
       end
     end
+
+    trait :with_values do
+      transient do
+        value_count { 5 }
```
Comment (skateman) [inline spec/factories/profile.rb:26]: The `value_count` variable renders the `with_values` obsolete ... please eliminate the `trait` similarly to rulegroups.

File: spec/factories/profile.rb [hunk 26]
```diff
@@ -20,5 +20,38 @@
...
+          object[value.ref_id] = SecureRandom.random_number(10)
+        end
+        profile.update!(value_overrides: values)
+      end
+    end
+
+    trait :with_rules do
+      transient do
+        rule_count { 5 }
+      end
```
Comment (skateman) [inline spec/factories/profile.rb:45]: Same here, unnecessary trait.

File: spec/factories/profile_rule.rb [hunk 27]
```diff
@@ -2,5 +2,7 @@
 
 FactoryBot.define do
   factory :v2_profile_rule, class: 'V2::ProfileRule' do
+    profile { association :v2_profile }
+    rule { association :v2_rule }
```
Comment (skateman) [inline spec/factories/profile_rule.rb:6]: Do we really need these default values?

File: spec/factories/security_guide.rb [hunk 28]
```diff
@@ -24,6 +24,20 @@
       end
     end
 
+    trait :with_rules do
+      transient do
+        rule_count { 3 }
+      end
```
Comment (skateman) [inline spec/factories/security_guide.rb:30]: Trait can be eliminated.

File: spec/factories/tailoring.rb [hunk 29]
```diff
@@ -0,0 +1,55 @@
...
+    trait :with_default_rules do
+      after(:create) do |tailoring, _|
+        tailoring.profile.rules.each do |rule|
+          FactoryBot.create(
+            :v2_tailoring_rule,
+            rule: rule,
+            tailoring: tailoring
+          )
+        end
+      end
```
Comment (skateman) [inline spec/factories/tailoring.rb:52]: I think this should be the default behavior, and maybe there could be the `with_no_rules` trait as an edge case. Most users have default tailoring anyway so it makes sense to me. WDYT?

File: spec/policies/v2/tailoring_policy_spec.rb [hunk 30]
```diff
@@ -0,0 +1,53 @@
...
+      :v2_tailoring,
+      policy: FactoryBot.create(
+        :v2_policy,
+        account: FactoryBot.create(:v2_user).account,
+        profile: canonical_profile
+      ),
+      profile: canonical_profile,
+      os_minor_version: 8
+    )
+  end
```
Comment (skateman) [inline spec/policies/v2/tailoring_policy_spec.rb:39]: Why is this here, what's its purpose?

**Review by skateman — CHANGES_REQUESTED** [review]
Please also revisit the controller specs for duplication of contexts, there are a lot of factories and some might be pulled out to higher contexts to deduplicate some code.

File: spec/policies/v2/tailoring_policy_spec.rb (hunk 30 above)
Comment (romanblanco) [reply spec/policies/v2/tailoring_policy_spec.rb:39]: The tailoring is used in the test that a user can't access foreign tailoring. 
Notice the account is changed (not `user.account`) in this instance.

File: spec/factories/tailoring.rb (hunk 29 above)
Comment (romanblanco) [reply spec/factories/tailoring.rb:52]: Makes sense to me. Default rules should be the default here.

File: spec/factories/profile.rb [hunk 31]
```diff
@@ -20,5 +20,34 @@
...
+      after(:create) do |profile, evaluator|
+        create_list(
+          :v2_value_definition,
+          evaluator.value_count,
+          security_guide: profile.security_guide
+        )
+        values = profile.security_guide.value_definitions.each_with_object({}) do |value, object|
+          object[value.ref_id] = SecureRandom.random_number(10)
+        end
+        profile.update!(value_overrides: values)
```
Comment (skateman) [inline spec/factories/profile.rb:38]: Maybe don't set all the definition values from a security guide, just a subset...same as not all rules are in a profile.

File: spec/factories/tailoring.rb [hunk 32]
```diff
@@ -0,0 +1,59 @@
...
+    end
+
+    trait :with_tailored_values do
+      value_overrides do
+        profile.value_overrides&.each do |k, v|
+          # overriding (tailoring) original values
+          profile.value_overrides[k] = v + SecureRandom.random_number(100)
+        end
+      end
+    end
```
Comment (skateman) [inline spec/factories/tailoring.rb:29]: Same as above, randomization of the right side isn't enough, it should be also just overriding a few values + shuffle in some from the security guide (follow what happens with rules).

**Review by skateman — APPROVED** [review]
The Seal of Approval
